### PR TITLE
AddressUtil normalize methods

### PR DESCRIPTION
**Pull request by blakegreen32**
Added normalizeRegion and normalizeCountry to AddressUtil

Ticket: https://nanopay.atlassian.net/browse/NP-5285

Nanopay PR: https://github.com/nanoPayinc/NANOPAY/pull/13312

### CODE REVIEW

File: src/foam/util/AddressUtil.java [hunk 1]
```diff
@@ -57,4 +66,42 @@
     return new String[] { streetNumber, streetName };
   }
 
+  public static String normalizeRegion(X x, String country, String regionCode) {
+    String normalizedRegion = country + "-" + regionCode;
+    DAO regionDAO = (DAO) x.get("regionDAO");
+    ArraySink sink = (ArraySink) regionDAO.where(AND(
```
Comment (jlhughes) [inline src/foam/util/AddressUtil.java]: why the use of AND?

Comment (jlhughes) [reply src/foam/util/AddressUtil.java]: And if you are testing for isocode or name below, why not add that to the above?
AND ( country ,
  OR ( iso,
          name

File: src/foam/util/AddressUtil.java [hunk 2]
```diff
@@ -57,4 +66,42 @@
...
+        break;
+      }
+    }
+    return normalizedRegion;
+  }
+
+  public static String normalizeCountry(X x, String countryCode) {
+    String normalizedCountry = countryCode;
+    DAO countryDAO = (DAO) x.get("countryDAO");
+    ArraySink sink = (ArraySink) countryDAO.select(new ArraySink());
```
Comment (jlhughes) [inline src/foam/util/AddressUtil.java]: Realize countryDAO is small but why not an OR mlang?

File: src/foam/util/AddressUtil.java (hunk 1 above)
Comment (nanoscott) [reply src/foam/util/AddressUtil.java]: updated to use CONTIANS_IC, which is our only option for ignores case

File: src/foam/util/AddressUtil.java (hunk 2 above)
Comment (nanoscott) [reply src/foam/util/AddressUtil.java]: also updated to use CONTAINS_IC

File: src/foam/util/AddressUtil.java [hunk 3]
```diff
@@ -57,4 +66,73 @@
...
+    String normalizedCountry = countryCode;
+    DAO countryDAO = (DAO) x.get("countryDAO");
+    Country found = (Country) countryDAO.find(countryCode);
+    if ( found != null )
+      return found.getCode();
+
+    ArraySink sink = (ArraySink) countryDAO
+      .where(OR(
+        EQ(Country.ISO31661CODE, countryCode),
+        CONTAINS_IC(Country.NAME, countryCode)
```
Comment (lchanmann) [inline src/foam/util/AddressUtil.java]: Might not be accurate when doing contain search? eg. searching for "United", should it be UK or US?

File: src/foam/util/AddressUtil.java [hunk 4]
```diff
@@ -57,4 +66,73 @@
...
+    DAO regionDAO = (DAO) x.get("regionDAO");
+    Region found = (Region) regionDAO.find(regionCode);
+    if ( found != null )
+      return found.getCode();
+
+    ArraySink sink = (ArraySink) regionDAO.where(AND(
+      EQ(Region.COUNTRY_ID, country),
+      OR(
+        EQ(Region.ISO_CODE, regionCode),
+        CONTAINS_IC(Region.NAME, regionCode)
```
Comment (lchanmann) [inline src/foam/util/AddressUtil.java]: May use starts_with_ic?

File: src/foam/util/AddressUtil.java (hunk 3 above)
Comment (nanoscott) [reply src/foam/util/AddressUtil.java]: country code should never be partial .. it should either be the whole country name, or either 3 digit ISO code or the 2 digit ISO code, so we should never get "United", only "United States" or "United Kingdom". The problem comes when we get something in IND and it matches Indonesia. In this case the logic below would kick in and the exact match on 3 digit ISO code would properly return India (IN) instead of Indonesia (ID)

File: src/foam/util/AddressUtil.java (hunk 4 above)
Comment (nanoscott) [reply src/foam/util/AddressUtil.java]: STARTS_WITH_IC does not have a Java implementation, only a javascript implementation.

Comment (nanoscott) [reply src/foam/util/AddressUtil.java]: Looks like the MLang.java implementation of STARTS_WITH_IC does not exist. I'll try to add it

Comment (nanoscott) [reply src/foam/util/AddressUtil.java]: Updated both CONTAINS_IC to STARTS_WITH_IC.

File: src/foam/util/AddressUtil.java (hunk 3 above)
Comment (lchanmann) [reply src/foam/util/AddressUtil.java]: assuming we never get "United" but if we do, should we throw instead of return US or UK?

Comment (nanoscott) [reply src/foam/util/AddressUtil.java]: I think a best effort match is better than entering an invalid country or throwing and not being able to save the address at all ...

Comment (lchanmann) [reply src/foam/util/AddressUtil.java]: Ok if we decided so, can we add documentation for future reference?

**Review by nanoscott — APPROVED** [review]

File: src/foam/util/AddressUtil.java [hunk 5]
```diff
@@ -57,4 +66,73 @@
...
+  }
+
+  public static String normalizeCountry(X x, String countryCode) {
+    String normalizedCountry = countryCode;
+    DAO countryDAO = (DAO) x.get("countryDAO");
+    Country found = (Country) countryDAO.find(countryCode);
+    if ( found != null )
+      return found.getCode();
+
+    ArraySink sink = (ArraySink) countryDAO
```
Comment (kgrgreer) [inline src/foam/util/AddressUtil.java]: Rather than selecting into an ArraySink, you could just process directly. Then you can stop processing once you find the correct value and you don't generate garbage.
See the following example taken from Group.js:

 .select(new AbstractSink() {
              public void put(Object o, Detachable d) {
                GroupPermissionJunction j = (GroupPermissionJunction) o;

                if ( j.getTargetId().isBlank() ) return;

                if ( j.getTargetId().startsWith("@") ) {
                  Group group = (Group) dao.find(j.getTargetId().substring(1));

                  if ( group != null && group.implies(x, permission) ) {
                    granted[0] = true;
                    d.detach();
                  }
                } else if ( new AuthPermission(j.getTargetId()).implies(permission) ) {
                  granted[0] = true;
                  d.detach();
                }
              }
            });

The d.detach() stops the select() early.
